README.md
# abalone
A simple Sinatra & hterm based web terminal.

#### Table of Contents

1. [Overview](#overview)
1. [Configuration](#configuration)
    1. [SSH](#configuring-ssh)
    1. [Custom Login Command](#configuring-a-custom-command)
1. [jQuery plugin](#jquery-plugin)
1. [Limitations](#limitations)

## Overview

Simply exposes a login shell to a web browser. This is currently
nowhere near to production quality, so don't actually use it.

It supports three methods for providing a shell:

1. Simply running the `login` binary and logging in as a system user. (default)
1. Using SSH to connect to localhost or a remote machine. This can be configured
   with credentials to automatically connect, or it can request a username and
   password from the end user.
1. Running custom command, which can be configured with arbitrary parameters.

## Configuration

Abalone defaults to loading configuration from `/etc/abalone/config.yaml`. You
can pass the path to another config file at the command line. In that file, you
can set several options:

* `:autoconnect`
  * Set this to true if you'd like the session to start on page load and false
    if you'd like the user to click a *Start Session* button instead. Defaults
    to `true`.
* `:port`
  * Which port to run the server on.
  * Default value: `9000`
* `:bind`
  * The hostname or IP address of the interface to listen on.
  * Default value: `0.0.0.0` (listen to all interfaces.)
* `:bannerfile`
  * File to display before login. This does not interpret special characters the way `getty` does.
  * `true`, `false`, or filename to display.
  * Default value: `false`, or `/etc/issue.net` if set to `true`.
* `:welcome`
  * A message to display prior to starting a session. This is on the overlay with the
    *Start Session* button. Pass a string of text, or a filename. HTML will be interpreted.
  * Default value: unset
* `:logfile`
  * The path of a file to log to.
  * Default value: Log only to `STDERR`. If you pass `-l` at the command line
    with no filename, it will log to `/var/log/abalone`.
* `:timeout`
  * Maximum number of seconds a session can last. The shell will be killed at the
    end of that time. For example, set it to 300 for shells that last for up to
    five minutes.
  * Default value: unset.
* `:ttl`
  * The number of seconds a session should last after disconnecting. If you reconnect
    within this grace period, you'll be reconnected to your session without interruption.
    This cannot yet restore the secondary terminal buffer, so if you're running something
    like Vim, you may have to run `clear` or `reset` after exiting to get your console
    sane again.
  * Note that `:timeout` takes precedence, so if your session times out, even during
    the `:ttl` grace period, it will be killed. 
  * Default value: unset.
* One of [`:command`](#configuring-a-custom-command) or [`:ssh`](#configuring-ssh), exclusive.
  * The login method to use. Abalone can use `login`, SSH, or a custom command
    to start a shell. See configuration instructions below.
  * Default value: uses the `login` binary, with no configuration possible.

### Configuring SSH

The following parameters may be used to configure SSH login. The `:host` setting
is required. `:user` and `:cert` are optional. If `:user` is not set, then the
user will be prompted for a login name, and if `:cert` is not set then
the user will be prompted to log in with a password. If the SSH server is running
on a non-standard port, you may specify that with the `:port` setting.

``` Yaml
---
:ssh:
  :host: shellserver.example.com
  :user: centos
  :cert: /etc/abalone/centos.pem
```

### Configuring a custom command

A custom command can be configured in several ways. If you just want to run a
command without providing any options, the config file would look like:

``` Yaml
---
:command: /usr/local/bin/run-container
```

#### Simple options

You can also allow the user to pass in a arbitrary options. These must be
whitelisted. You can simply list allowed options in an Array:

``` Yaml
---
:command: /usr/local/bin/run-container
:params: [ 'username', 'image' ]
```

The options will be passed to the command in this way, ignoring the option
that was not whitelisted:

* http://localhost:9000/?username=bob&image=testing&invalid=value
* `/usr/local/bin/run-container --username bob --image testing`

#### Customized options

Finally, you can fully customize the options which may be passed in, including
remapping them to command line arguments and filtering accepted values. In this
case, `:params` must be a Hash.

``` Yaml
---
:command: /usr/local/bin/run-container
:params:
  username:
  type:
    :values: ['demo', 'testing']
  image:
    :map: '--create-image'
    :values: [ 'ubuntu', 'rhel', /centos[5,6,7]/ ]
```

Notice that `username` has nothing on the right side. It will be treated exactly
the same as `username` in the *Simple options* array above.

The `image` parameter is more complex though. It has two keys specified. Both are
optional. If `:map` is set, then its value will be used when running the command.
The `:values` key can be used to specify a list of valid values. Note that these
can be specified as Strings or regular expressions. 

The options in this case will be passed to the command like:

* An option is mapped to a different command line argument:
  * http://localhost:9000/?username=bob&image=centos6
  * `/usr/local/bin/run-container --username bob --create-image centos6`
* An option without a `:map` is passed directly through to the command:
  * http://localhost:9000/?username=bob&type=demo
  * `/usr/local/bin/run-container --username bob --type demo`
* Image name that doesn't pass validation is ignored:
  * http://localhost:9000/?username=bob&image=testing
  * `/usr/local/bin/run-container --username bob`
* Invalid options and values are ignored:
  * http://localhost:9000/?username=bob&image=testing&invalid=value
  * `/usr/local/bin/run-container --username bob`

## jQuery Plugin

Abalone comes with a build in jQuery plugin that makes it very easy to use. You
can attach the launcher to any element. If it's a `block` element, then a launcher
button will be injected inside, and if it's `inline` then it will directly trigger
the terminal.

See a demo of the launcher in action after installation by starting the server and
browsing to [http://localhost:9000/demo.html](http://localhost:9000/demo.html).
Adjust the URL and port as needed.

The minimum external dependencies are jQuery and jQuery UI:

    <link rel="stylesheet" href="https://code.jquery.com/ui/1.12.1/themes/base/jquery-ui.css">
    <script src="https://code.jquery.com/jquery-1.12.4.min.js"></script>
    <script src="https://code.jquery.com/ui/1.12.1/jquery-ui.js"></script>

To load and initialize the launcher, you'll need to load the CSS and Javascript
from a running Abalone instance like below. Notice the full URL, including the
port number. Alternatively, you can pull those from the repository and host them
along with the rest of your HTML.

    <link rel="stylesheet" href="http://localhost:9000/css/launcher.css">
    <script src="http://localhost:9000/js/launcher.js"></script>

Then you'll simply declare one or more launchers on any element you choose. Note
that you must pass in the server parameter. This should be the location of your
Abalone server, including the port it's running on.

    <script>
      $(document).ready(function() {

        $('pre.popup').AbaloneLauncher({
           label: "Try out a popup!",
           title: "Isn't this neat?",
          server: "http://localhost:9000",
        });

        $('pre.inline').AbaloneLauncher({
             label: "Try it out inline!",
            target: "inline",
            server: "http://localhost:9000",
        });

        $('pre.targeted').AbaloneLauncher({
             label: "Try it out!",
            target: "#abalone-shell",
          location: "se",
            server: "http://localhost:9000",
        });

        $('a#launcher').AbaloneLauncher({
          server: "http://localhost:9000",
          params: { "type": "demo", "uuid": generateUUID() },
        });

      });
    </script>

### Configuration Options

| Option     | Valid values                                          | Default               |
|------------|-------------------------------------------------------|-----------------------|
| `location` | `ne`, `se`, `sw`, `nw`                                | `ne`                  |
| `label`    | *String*                                              | *Launch*              |
| `title`    | *String*                                              | *Abalone Web Shell*   |
| `target`   | `popup`, `inline`, `tab`, CSS selector of a container | `popup`               |
| `params`   | parameters to be passed to the server                 | `{}`                  |
| `server`   | URL to the Abalone server, including port             | `null` (**required**) |
| `height`   | *Integer*                                             | `480`                 |
| `width`    | *Integer*                                             | `640`                 |


## Limitations

This is super early in development and has not yet been battle tested.

## Disclaimer

I take no liability for the use of this tool.

Contact
-------

binford2k@gmail.com



CHANGELOG.md
# v0.4.1

* Corrected param handling when defined as an array in the `config.yaml`.

# v0.4.0

* Add a `:ttl` setting, which lets you reconnect to a running session for a
  given number of seconds after disconnecting. This allows you to resume sessions
  interrupted by network instability, for example.
* Add a menu option to force a terminal reset, e.g. killing & restarting.
* Add a jQuery plugin which makes it trivial to use Abalone in an existing page.
* Add a `:welcome` setting, which puts a message on the *Start Session* overlay.

# v0.3.4

* Added a watchdog when running under systemd.

# v0.3.3

* Added a new `bannerfile` setting.
* Set to:
    * filename to display
    * `false`: no file is displayed
    * `true: display `/etc/issue.net`


# v0.3.2

* Disable X-Frame-Options so this can be easily embedded.

# v0.3.1

* Ensures the environment is somewhat sane.

# v0.3.0

* Adds a timeout counter, allowing a maximum life setting for shells.

# v0.2.1

* Fixed a string encoding issue that could sometimes cause wedged shells.

# v0.2.0

* Added autoconnect setting so you can require users click
  a button to start their session.

# v0.1.1

* Fixed a crashing UTF8 bug
* Added a simple Puppet module for management

# v0.1.0

* Added customization options and documentation.
* Can now run `login`, `ssh`, or a custom command.
* Can use an SSH key for passwordless login.
* Can accept arbitrary whitelisted command arguments.
  * passed to command via parameters, so no shell injection.

# v0.0.1

* Initial release.


lib/abalone.rb
require 'json'
require 'logger'
require 'sinatra/base'
require 'sinatra-websocket'

class Abalone < Sinatra::Base
  set :logging, true
  set :strict, true
  set :protection, :except => :frame_options
  set :public_folder, "#{settings.root}/../public"
  set :views, "#{settings.root}/../views"
  set :active, {}

  enable :sessions

  before {
    env["rack.logger"] = settings.logger if settings.logger

    trap('INT') do
      # this forces all the spawned children to terminate as well
      puts "Caught SIGINT. Terminating active sessions (#{Process.pid}) now."
      exit!
    end
  }

  # super low cost heartbeat response.
  # Add this first to ensure that the user route doesn't take precedence.
  get '/heartbeat/ping' do
    'alive'
  end

  get '/api/reset' do
    if @terminal = settings.active[session.id]
      @terminal.terminate!
    end
    redirect to('/')
  end

  get '/?:user?' do
    if !request.websocket?
      @requestUsername = (settings.respond_to?(:ssh) and ! settings.ssh.include?(:user)) rescue false
      @autoconnect     = settings.autoconnect
      if settings.respond_to?(:welcome)
        begin
          @welcome = File.file?(settings.welcome) ? File.read(settings.welcome) : settings.welcome
        rescue => e
          warn e.message
        end
      end
      erb :index
    else
      request.websocket do |ws|
        uid = session.id
        ws.onopen do
          warn("websocket opened")
          # expire all dead terminals
          settings.active.delete_if {|uid, term| ! term.alive? }

          if @terminal = settings.active[uid]
            @terminal.reconnect(ws)
          else
            warn "Starting a new session for #{uid}."
            @terminal = Abalone::Terminal.new(settings, ws, sanitized(params))

            if settings.respond_to? :ttl
              settings.active[uid] = @terminal
              logger.debug "Saving session #{uid}."
              logger.debug "Active sessions: #{settings.active.keys.inspect}"
            end
          end
        end

        ws.onclose do
          warn('websocket closed')
          @terminal.stop! if @terminal
        end

        ws.onmessage do |message|
          message = JSON.parse(message)

          begin
            case message['event']
            when 'input'
              @terminal.write(message['data'])

            when 'modes'
              @terminal.modes = message['data']

            when 'resize'
              row = message['row']
              col = message['col']
              @terminal.resize(row, col)

            when 'logout', 'disconnect'
              warn("Client exited.")
              @terminal.stop!

            else
              warn("Unrecognized message: #{message.inspect}")
            end
          rescue Errno::EIO => e
            warn "Remote terminal closed."
            warn e.message
            @terminal.stop!

          end

        end
      end
    end
  end

  not_found do
    halt 404, "You shall not pass! (page not found)\n"
  end

  helpers do

    def sanitized(params)
      params.reject do |key,val|
        ['captures','splat'].include?(key) or not allowed(key, val)
      end
    end

    def allowed(param, value)
      return false unless settings.params.include? param
      return true if settings.params.is_a? Array

      config = settings.params[param]
      return true if config.nil?
      return true unless config.include? :values

      config[:values].each do |pattern|
        case pattern
        when String
          return true if value == pattern
        when Regexp
          return true if pattern.match(value)
        end
      end

      false
    end

  end
end


public/js/abalone.js
var term;
var buf = '';
var socket;

function connect(userPrompt) {
  var protocol = (location.protocol === 'https:') ? 'wss://' : 'ws://';

  // hide the connect button
  document.getElementById("overlay").style.display = "none";

  if(socket) {
    socket.onclose = null;
    socket.close();
  }

  if(userPrompt && location.pathname == '/') {
    var username = prompt('What username would you like to connect with?');
    socket = new WebSocket(protocol + location.host + '/' + username + location.search);
  }
  else {
    socket = new WebSocket(protocol + location.host + location.pathname + location.search);
  }

  socket.onopen    = function()  { connected();            };
  socket.onclose   = function()  { disconnected();         }
  socket.onmessage = function(m) { messageHandler(m.data); };
}

function connected() {
  console.log('Client connected');
  lib.init(function() {
      hterm.defaultStorage = new lib.Storage.Local();
      term = new hterm.Terminal();
      window.term = term;
      term.decorate(document.getElementById('terminal'));

      term.setCursorPosition(0, 0);
      term.setCursorVisible(true);
      term.prefs_.set('ctrl-c-copy', true);
      term.prefs_.set('ctrl-v-paste', true);
      term.prefs_.set('use-default-window-copy', true);

      term.runCommandClass(Abalone, document.location.hash.substr(1));
      socket.send(JSON.stringify({
        event: 'resize',
        col: term.screenSize.width,
        row: term.screenSize.height
      }));

      if (buf && buf != '')
      {
          term.io.writeUTF16(buf);
          buf = '';
      }
  });

  /* save our terminal state on the server periodically, just in case we restart a session */
  // Note: Don't try to tie it to the setDECMode() method as that's called all the time.
  setInterval(function() {
    socket.send(JSON.stringify({
      event: 'modes',
       data: getModes()
    }));
  }, 5000);

}

function disconnected() {
  console.log('Client disconnected');
  document.getElementById("overlay").style.display = "block";
}

function messageHandler(message) {
  var message = JSON.parse(message);
  var event = message['event'];
  var data  = message['data'];

  switch(event) {
    case 'time':
      document.getElementById("timer").style.display = "block";
      document.getElementById("timer").innerHTML = data;
      break;

    case 'modes':
      setModes(data);

    default:
      if (!term) {
          buf += data;
          return;
      }
      term.io.writeUTF16(data);
  }
}

function getModes() {
  return {
    "cursorBlink"       : term.vt.terminal.options_.cursorBlink,
    "cursorVisible"     : term.vt.terminal.options_.cursorVisible,
    "bracketedPaste"    : term.vt.terminal.options_.bracketedPaste,
    "applicationCursor" : term.vt.terminal.keyboard.applicationCursor
  };
}

function setModes(modes) {
  term.vt.terminal.options_.cursorBlink       = modes["cursorBlink"];
  term.vt.terminal.options_.cursorVisible     = modes["cursorVisible"];
  term.vt.terminal.options_.bracketedPaste    = modes["bracketedPaste"];
  term.vt.terminal.keyboard.applicationCursor = modes["applicationCursor"];
}

/* borrowed from https://github.com/krishnasrinivas/wetty */
function Abalone(argv) {
  this.argv_ = argv;
  this.io = null;
  this.pid_ = -1;
}

Abalone.prototype.run = function() {
  this.io = this.argv_.io.push();

  this.io.onVTKeystroke = this.sendString_.bind(this);
  this.io.sendString = this.sendString_.bind(this);
  this.io.onTerminalResize = this.onTerminalResize.bind(this);
}

Abalone.prototype.sendString_ = function(str) {
  socket.send(JSON.stringify({ event: 'input', data: str}));
};

Abalone.prototype.onTerminalResize = function(col, row) {
  socket.send(JSON.stringify({ event: 'resize', col: col, row: row}));
};



lib/abalone/terminal.rb
require 'pty'
require 'io/console'
require 'abalone/buffer'

class Abalone::Terminal
  def initialize(settings, ws, params)
    @settings = settings
    @ws       = ws
    @params   = params
    @modes    = nil
    @buffer   = Abalone::Buffer.new

    ENV['TERM'] ||= 'xterm-256color' # make sure we've got a somewhat sane environment

    if settings.respond_to?(:bannerfile)
      @ws.send({'data' => File.read(settings.bannerfile).encode(crlf_newline: true)}.to_json)
      @ws.send({'data' => "\r\n\r\n"}.to_json)
    end

    reader, @writer, @pid = PTY.spawn(*shell_command)
    @writer.winsize = [24,80]

    # there must be some form of event driven pty interaction, EM or some gem maybe?
    reader.sync = true
    @term = Thread.new do
      carry = []
      loop do
        begin
          PTY.check(@pid, true)
          output = reader.read_nonblock(512).unpack('C*') # we read non-blocking to stream data as quickly as we can
          last_low = output.rindex { |x| x < 128 } # find the last low bit
          trailing = last_low +1

          # use inclusive slices here
          data  = (carry + output[0..last_low]).pack('C*').force_encoding('UTF-8') # repack into a string up until the last low bit
          carry = output[trailing..-1]             # save the any remaining high bits and partial chars for next go-round

          @ws.send({'data' => data}.to_json)

        rescue IO::WaitReadable
          IO.select([reader])
          retry

        rescue PTY::ChildExited => e
          warn('Terminal has exited!')
          @ws.close_connection

          @timer.terminate rescue nil
          @timer.join rescue nil
          Thread.exit
        end

        sleep(0.05)
      end
    end

    if @settings.respond_to? :timeout
      @timer = Thread.new do
        expiration = Time.now + @settings.timeout
        loop do
          remaining = expiration - Time.now
          if remaining < 0
            terminate!
            Thread.exit
          end

          format = (remaining > 3600) ? "%H:%M:%S" : "%M:%S"
          time = {
            'event' => 'time',
            'data'  => Time.at(remaining).utc.strftime(format),
          }
          @ws.send(time.to_json)
          sleep 1
        end
      end
    end
  end

  def alive?
    @term.alive?
  end

  def reconnect(ws)
    if @ttl # stop the countdown
      warn "Stopping timeout"
      @ttl.terminate rescue nil
      @ttl.join rescue nil
      @ttl = nil
    end
    @ws.close_connection if @ws
    @ws = ws

    sleep 0.25 # allow the terminal to finish initialization before we blast it.

    if @modes
      @ws.send({
        'event' => 'modes',
        'data'  => @modes
      }.to_json)
    end
    @ws.send({'data' => @buffer.replay}.to_json)
    @writer.write "\cl" # ctrl-l forces a screen redraw
  end

  def write(message)
    @writer.write message
  end

  def modes=(message)
    raise 'Invalid modes data type' unless message.is_a? Hash
    @modes = message.select do |key, val|
      ['cursorBlink', 'cursorVisible', 'bracketedPaste', 'applicationCursor'].include? key
    end
  end

  def stop!
    if @settings.respond_to? :ttl
      @ws  = @buffer
      @ttl = Thread.new do
        warn "Providing a shutdown grace period of #{@settings.ttl} seconds."
        sleep @settings.ttl
        terminate!
      end
    else
      terminate!
    end
  end

  def terminate!
    warn "Terminating session."
    Process.kill('TERM', @pid) rescue nil
    sleep 1
    Process.kill('KILL', @pid) rescue nil

    [@ttl, @timer, @term].each do |thread|
      thread.terminate rescue nil
      thread.join rescue nil
    end
  end

  def resize(rows, cols)
    @writer.winsize = [rows, cols]
  end

  private
  def shell_command
    if @settings.respond_to? :command
      return @settings.command unless @settings.respond_to? :params

      command = @settings.command
      command = command.split if command.class == String

      @params.each do |param, value|
        if @settings.params.is_a? Array
          command << "--#{param}" << value
        else
          config = @settings.params[param]
          case config
          when nil
            command << "--#{param}" << value
          when Hash
            command << (config[:map] || "--#{param}")
            command << value
          end
        end
      end

      return command
    end

    if @settings.respond_to? :ssh
      config = @settings.ssh.dup
      config[:user] ||= @params['user'] # if not in the config file, it must come from the user

      if config[:user].nil?
        warn "SSH configuration must include the user"
        return ['echo', 'no username provided']
      end

      command = ['ssh', config[:host] ]
      command << '-l' << config[:user] if config.include? :user
      command << '-p' << config[:port] if config.include? :port
      command << '-i' << config[:cert] if config.include? :cert

      return command
    end

    # default just to running login
    'login'
  end

end

lib/abalone/watchdog.rb
class Abalone::Watchdog
  def initialize(period, options)
    raise 'Watchdog needs a port to be set' unless options.include? :port
    raise 'Watchdog needs a logger to be set' unless options.include? :logger
    logger = options[:logger]

    Thread.new do
      require 'net/http'
      sleep 5 # give the service time to wake up

      logger.warn "Starting watchdog with period #{period}"
      loop do
        http = Net::HTTP.new('localhost', options[:port])
        http.open_timeout = 1
        http.read_timeout = 1
        begin
          http.start
          http.request_get('/heartbeat/ping') do |res|
            logger.debug "Heartbeat response: #{res.read_body}"
            system('systemd-notify WATCHDOG=1')
          end
        rescue => e
          logger.warn 'Abalone service failed heartbeat check!'
          logger.debug e.message
        end

        sleep period
      end
    end
  end
end


public/js/launcher.js
(function ( $ ) {
    /* global counter for constructing unique IDs */
    abaloneInstanceCount = 0;

    $.fn.AbaloneLauncher = function(options) {
      var settings = $.extend({
      location: 'ne',
         label: "Launch",
         title: "Abalone Web Shell",
        target: "popup",
        params: {},
        server: null,
        height: 480,
         width: 640,
      }, options );

      return this.each(function() {
        if (settings.server == null) {
          console.log("[FATAL] Abalone: server is a required parameter.")
          return;
        }

        abaloneInstanceCount++;
        var element   = $(this);
        var serverURL = settings.server + '?' + $.param(settings.params);

        if (element.css("display") == "block") {
          var launcher = $("<input>", {
                             "type": "button",
                            "class": "launcher",
                            "value": settings.label,
                          });

          launcher.addClass("control");

          if (-1 != $.inArray(settings.location, ["ne", "se", "sw", "nw"])) {
            launcher.addClass("location-" + settings.location);
          }

          /* we need this for the absolutely posititioned button */
          if (element.css("position") == "static" ) {
            element.css("position", "relative");
          }

          element.prepend(launcher);
        } else {
          if (settings.target == "inline") {
            console.log("[FATAL] Abalone: you cannot use the inline target without a container.")
            return;
          }
          var launcher = element;
        }
        launcher.addClass("abalone launcher instance-"+abaloneInstanceCount);
        launcher.click(function(e) {
          e.preventDefault();
          var button = $(this);
          switch(settings.target) {
            case "popup":
              var abalone = $("<iframe>", { "class": "abalone popup", "src": serverURL });
              abalone.dialog({
                height: settings.height,
                 width: settings.width,
                 title: settings.title,
                 close: function( event, ui ) {
                          button.prop('disabled', false);
                          abalone.remove();
                        }
              });
              button.prop('disabled', true);
              break;

            case "inline":
              var abalone = $("<iframe>", { "class": "abalone inline", "src": serverURL });
              element.append(abalone);
              break;

            case "tab":
              window.open(serverURL, 'abaloneTerminal');
              break

            /* Assume that the user has passed in a string as a selector target */
            default:
              var abalone = $("<iframe>", { "class": "abalone targeted", "src": serverURL });
              var target  = $(settings.target)
              target.append(abalone);
              break;
          }

          /* swap out for the close button, unless we're using a popup/tab */
          if (["popup", "tab"].indexOf(settings.target) == -1) {
            var close = $("<input>", {
               "type": "button",
              "class": "abalone inline control exit location-" + settings.location,
              "value": "Close",
            });
            close.on("click", function() {
              $(this).remove();
              abalone.remove();
              button.show();
            });

            element.prepend(close);
            button.hide();
          }

        });

        return this;
      });
    };

}(jQuery));
